sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+/* Handles DTLS socket events from the event queue */
+static void _on_sock_dtls_evt(sock_dtls_t *sock, sock_async_flags_t type, void *arg) {
+    (void)arg;
+    coap_socket_t socket = { .type = COAP_SOCKET_TYPE_DTLS, .socket.dtls = sock};
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &socket.ctx_dtls_session,
+                            _listen_buf, sizeof(_listen_buf),
+                            CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+        if (res != -SOCK_DTLS_HANDSHAKE) {
+            DEBUG("gcoap: could not establish DTLS session: %d\n", res);
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+            return;
+        }
+        dsm_state_t prev_state = dsm_store(sock, &socket.ctx_dtls_session,
+                                           SESSION_STATE_ESTABLISHED, false);
+
+        /* If session is already stored and the state was SESSION_STATE_HANDSHAKE
+        before, the handshake has been initiated internally by a gcoap client request
+        and another thread is waiting for the handshake. Send message to the
+        waiting thread to inform about established session */
+        if (prev_state == SESSION_STATE_HANDSHAKE) {
+            msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+            msg_send(&msg, _auth_waiting_thread);
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+        }
+
+        /* If not enough session slots left: set timeout to free up session */
+        uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+        if (dsm_get_num_available_slots() < minimum_free)
+        {
+            uint32_t timeout = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS_TIMEOUT_USEC;
+            event_callback_init(&_dtls_session_free_up_tmout_cb,
+                                _dtls_free_up_session, NULL);
+            event_timeout_init(&_dtls_session_free_up_tmout, &_queue,
+                               &_dtls_session_free_up_tmout_cb.super);
+            event_timeout_set(&_dtls_session_free_up_tmout, timeout);
+        }
+    }
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        if (sock_dtls_get_event_session(sock, &socket.ctx_dtls_session)) {
+            dsm_remove(sock, &socket.ctx_dtls_session);
+        } else {
+            puts("gcoap: A session was closed, but the corresponding session " \
+            "could not be retrieved from the socket!");
+            return;
+        }
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+
+        /* Remove all memos of the concerned session. TODO: oberservable memos! */
+        for (int i = 0; i < CONFIG_GCOAP_REQ_WAITING_MAX; i++) {
+            if (_coap_state.open_reqs[i].state == GCOAP_MEMO_UNUSED) {
```

Comment:
Do we need to lock `_coap_state.lock` here?

Reply:
Not 100% sure, from looking into code I didn't see locking when memos were set to unused. Since no moving or overwriting happens and no other* thread could access this memo I would say it's safe without.

* This path occurs when the remote peer closed the session, so we are already in the receiver thread. If context switch happens to a sending thread, we would only create a new memo and not edit existing ones.